### PR TITLE
fix: Make a ServerId type that's consistent everywhere

### PR DESCRIPTION
**Pull request by carols10cents**
Servers have an ID, which can only be set once per server instance and cannot be zero.

These IDs also get passed around with writes and replication and used in object storage.

The "set only once" constraint only needs to be enforced on the `Server` ID getter/setter methods, but the nonzero part should be enforced everywhere. There were a few places (in tests) that were using 0 for the concept of a server ID, which shouldn't be possible anywhere.

`ServerId` should be a separate type from `NonZeroU32` (rather than an alias) so that they're not interchangeable (and if we want to change to, say, a `NonZeroU64` inner type, we'll have one place to change it).

This does change the management service gRPC definitions, I think for the better, because some servers might never be writers but they still need an ID. So it makes more sense to set a server ID rather than a writer ID. We'll send a PR to conductor soon.

### CODE REVIEW

**Review by domodwyer — APPROVED** [review]
❤️  consistency - way to go!

File: data_types/src/server_id.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    fn eq(&self, other: &u32) -> bool {
+        self.get_u32() == *other
+    }
+}
+
+impl PartialEq<ServerId> for u32 {
+    fn eq(&self, other: &ServerId) -> bool {
+        *self == other.get_u32()
+    }
+}
```
Comment (crepererum) [inline data_types/src/server_id.rs:63]: This kinda "escapes" the type-safety that we get by having a dedicated type. Are they really required?

**Review by mkmik — APPROVED** [review]
I like server ID.

The fact that in practice we also use it so that each writer has their own "dedicated space", doesn't literally require calling this thing a "writer id"

File: data_types/src/server_id.rs (hunk 1 above)
Comment (carols10cents) [reply data_types/src/server_id.rs]: They made the tests a tiny bit nicer, but you're right that they're not really necessary and could actually be harmful. I pushed a commit taking them out!

**Pull request merged**